exporters/otlp/otlpgrpc/retry.go

```diff
@@ -0,0 +1,132 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package otlpgrpc
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/cenkalti/backoff/v4"
+	"google.golang.org/genproto/googleapis/rpc/errdetails"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	"go.opentelemetry.io/otel/exporters/otlp"
+)
+
+func doRequest(ctx context.Context, fn func(context.Context) error, rs otlp.RetrySettings, stopCh chan struct{}) error {
+	expBackoff := otlp.NewExponentialConfig(rs)
+
+	for {
+		err := fn(ctx)
+		if err == nil {
+			// request succeeded.
+			return nil
+		}
+
+		if !rs.Enabled {
+			return err
+		}
+
+		// We have an error, check gRPC status code.
+		st := status.Convert(err)
+		if st.Code() == codes.OK {
+			// Not really an error, still success.
+			return nil
+		}
+
+		// Now, this is this a real error.
+
+		if !shouldRetry(st.Code()) {
+			// It is not a retryable error, we should not retry.
+			return err
+		}
+
+		// Need to retry.
+		var delay time.Duration
+
+		// Respect server throttling.
+		if throttle := getThrottleDuration(st); throttle != 0 {
+			delay = throttle
+		} else {
+			backoffDelay := expBackoff.NextBackOff()
+			if backoffDelay == backoff.Stop {
+				// throw away the batch
+				err = fmt.Errorf("max elapsed time expired %w", err)
+				return err
+			}
+			delay = backoffDelay
+		}
```

Comment:
Do we want to pause the exponential backoff sequence if we get a throttle request in the middle of it, or should be progress the series as well as wait for the throttle request time?

Reply:
> Generally what is recommended is you take the max of InitalBackoff, LastBackoff*rate, and Threshold, apply jitter to that, and wait that long.

+1

Reply:
Using this `backoff` package we don't have control of the current interval, the jitter and etc.

I've changed to get the max between the server throttle and the next backoff wait time, is there an objection against that? @MrAlias @MadVikingGod

Reply:
SGTM :+1:

Reply:
Hmm, reading [the specification](https://github.com/open-telemetry/opentelemetry-specification/blob/b46bcab5fb709381f1fd52096a19541370c7d1b3/specification/protocol/exporter.md#retry), it looks like we need to add our own jitter to this to make sure we are compliant:

> This retry strategy MUST implement an exponential back-off with jitter to avoid overwhelming the destination until the network is restored or the destination has recovered.

Reply:
The current implementation does add a jitter when backing off, it's just a matter if the server sent us a `retry_delay` for us or not. 

Basically, if some transient error, we do exponential backoff with jitter but if the server send us `retry_delay` and that that is greater then the current backoff time (with jitter), we would honor that without adding jitter to the time that the server *asked* for us to wait.

I don't think we should add jitter to the retry delay that server send to us. It is responsibility of the server to give us a `retry_delay` that should be good enough to not overwhelm it.

Reply:
> I don't think we should add jitter to the retry delay that server send to us. It is responsibility of the server to give us a `retry_delay` that should be good enough to not overwhelm it.

Agreed, I think I misunderstood. This all sounds good :+1:

Reply:
>I don't think we should add jitter to the retry delay that server send to us. It is responsibility of the server to give us a retry_delay that should be good enough to not overwhelm it.

In reality, when a server is overwhelmed it has gotten too many requests at once, it will use a constant backoff so it doesn't get more overwhelmed figuring out how long clients really should wait, and if all the clients return at the same time you overwhelm it again.  That is why you add jitter to **ALL** backoffs, it's just when the server requests it there should be the minimum.

I think this is a good compromise.  If we add any more of the logic of how it "should" be done we are just reimplementing our dependency.